third_party/libunwindstack/MapInfo.cpp

```diff
@@ -176,6 +192,146 @@ Memory* MapInfo::GetFileMemory() {
+    if (map_info_it->name() == prev_name) {
+      if ((map_info_it->flags() & PROT_EXEC) != 0) {
+        // There is already an executable mapping for this name.
+        return nullptr;
+      }
+      first_map_info_with_prev_name = map_info_it;
+    }
+
+    map_info_it = map_info_it->prev_map();
+  }
+  CHECK(first_map_info_with_prev_name != nullptr)
+
+  if (first_map_info_with_prev_name->start() < first_map_info_with_prev_name->offset()) {
+    // base_address would be the result of an underflow. Even if this could theoretically happen, it
+    // is unexpected, so simply fail in this case.
+    return nullptr;
+  }
+
+  auto file_memory = std::make_unique<MemoryFileAtOffset>();
+  // Always use zero as offset because the headers of a PE should always be right at the beginning
+  // of the file.
+  if (!file_memory->Init(prev_name, 0)) {
+    return nullptr;
+  }
+
+  // Ideally we would use first_map_info_with_prev_name->GetObject(), but that's not right because
+  // it also requires an expected architecture. Since this function is called at most once for each
+  // map, we are fine with temporarily creating a new PeCoff.
+  PeCoff pe{file_memory.release()};
+  if (!pe.Init()) {
+    // The candidate file path doesn't correspond to a PE/COFF.
+    return nullptr;
+  }
+
+  // The address at which the first byte of the PE is mapped.
+  uint64_t base_address =
+      first_map_info_with_prev_name->start() - first_map_info_with_prev_name->offset();
+
+  uint64_t image_base_plus_text_virtual_address{};  // Image base + offset in memory.
+  uint64_t text_virtual_size{};                     // Size in memory.
+  if (!pe.GetTextRange(&image_base_plus_text_virtual_address, &text_virtual_size)) {
+    return nullptr;
+  }
+  const uint64_t text_virtual_address =
+      image_base_plus_text_virtual_address - pe.GetLoadBias();  // Offset in memory.
+
+  // The addresses at which the text section of the PE is supposed to be mapped.
+  const uint64_t expected_text_start = text_virtual_address + base_address;
+  const uint64_t expected_text_end = expected_text_start + text_virtual_size;
+
+  if (start() > expected_text_start || end() < expected_text_end) {
+    // This map doesn't fully contain the address range at which the text section of the PE is
+    // supposed to be mapped.
+    return nullptr;
+  }
+
+  auto memory = std::make_unique<MemoryFileAtOffset>();
+  // Always map the whole PE, as we always need the headers.
+  if (!memory->Init(prev_name, 0)) {
+    return nullptr;
+  }
+
+  // Compute the offset that this mapping would have if it was a file mapping (it's not a multiple
+  // of the page size, otherwise Wine would have created a file mapping to begin with).
+  const uint64_t text_pointer_to_raw_data = pe.GetTextOffsetInFile();
+  const uint64_t offset = text_pointer_to_raw_data - (expected_text_start - start());
+
+  Log::Info("Guessing that executable map at %#lx-%#lx belongs to \"%s\" at offset %#lx", start(),
```

Comment:
I am wondering if we should also add logging for the "unexpected" cases above where we early return?

Reply:
I considered this also in the PR tackling `orbit_object_utils::ParseMaps` and then decided against because some of the early returns can happen quite a bit, so we would have to choose carefully which early returns to log. It gets a bit confusing. In the end it's not really necessary. Also there is really not a lot of logging in libunwindstack and I want to keep it to a minimum here too.

Reply:
Thanks for the explanation. SGTM.